Привет! Меня зовут Исмаил, я буду проверять твой проект :) Можешь обращаться ко мне на «ты». Если тебе комфортно, то и я буду к тебе так обращаться, если нет, то обязательно скажи об этом.

Пожалуйста, не удаляй мои комментарии, которые я буду оставлять в работе. Это особенно поможет, если твои проекты будут отправлены на повторную проверку. При повторных проверках у комментариев будут приписки: "Вторая итерация", "Третья итерация" и т.д. 

Ты тоже можешь реагировать на мои комментарии, но в таком случае постарайся, чтобы твои комментарии отличались от моих: например, выделяй их своим любимым цветом — так у нас не возникнет путаницы :)

Кстати, про цвета! Мои комментарии будут в следующей цветовой гамме: зелеными, желтыми и красными. Например:

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
В случае, если решение на отдельном шаге является полностью правильным. </div>

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 
    
В случае, если решение может стать еще лучше с некоторыми корректировками </div>
 
<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 
    
В случае, если решение какого-то шага требует значительных изменений. Проект не может быть принят с первого раза, если ревью содержит комментарии, помеченные этим цветом </div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Прежде всего, посчитаем долю просрочивших для каждого варианта количества детей:

In [23]:
data.groupby('children')['debt'].mean()

children
0    0.075438
1    0.092346
2    0.094542
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

Отсутствие просрочивших среди дебиторов с пятью детьми, вероятно, объясняется тем, что их просто слишком мало в наборе данных, чтобы обнаружить хотя бы одного просрочившего. Проверим это предположение, добавиви в таблицу их точное количество:

In [29]:
data.groupby('children')['debt'].agg(['count', 'mean'])

count      mean
children                 
0         14091  0.075438
1          4808  0.092346
2          2052  0.094542
3           330  0.081818
4            41  0.097561
5             9  0.000000

Действительно, дебиторов с пятью детьми лишь девять человек, а в среднем на одного просрочившего приходится более десяти выплативших кредит вовремя.

**Вывод:** Судя по всему, на возврат кредита в срок влияет не количество детей, а их наличие. Так, среди тех, у кого нет детей, процент должников составлял порядка 7.5%, а среди остальных - от 8.2% до 9.8%.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
Здорово, что смотришь на относительные значения, на проценты. Также берешь во внимание кол-во данных о семьях с 5 детьми - по такому маленькому кол-ву действительно не стоит делать какие-то выводы, ибо выборка мала. Есть промежуточный комментарий/вывод и он верный. 

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 

Но хотелось бы, чтобы это была единая таблица: с кол-вом рядом с процентами

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Прежде всего, посчитаем долю просрочивших для каждого семейного статуса (чтобы выявить слишком малые категории, вновь добавим в таблицу общее количество дебиторов каждой категории):

In [32]:
data.groupby('family_status')['debt'].agg(['count', 'mean']).sort_values('mean', ascending=False)

count      mean
family_status                         
Не женат / не замужем   2796  0.097639
гражданский брак        4134  0.093130
женат / замужем        12261  0.075606
в разводе               1189  0.070648
вдовец / вдова           951  0.066246

**Вывод:** Процент просрочивших выше среди тех, кто никогда не состоял в официальном браке (неженатые и состоящие в гражданском браке) и составляет 9.3%-9.8%. Среди тех, кто когда-либо состоял в браке (женатые, вдовцы и разведённые) процент должников ниже и составляет 6.6%-7.6%.

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 

А тут почему бы не вывести кол-во? Может, будет такая же ситуация, что и в прошлом подпункте?

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Прежде всего, посчитаем долю просрочивших для каждой категории уровня доходов и общее количество дебиторов в ней:

In [30]:
data.groupby('total_income_category')['debt'].agg(['count', 'mean'])

count      mean
total_income_category                 
A                         25  0.080000
B                       5014  0.070602
C                      15921  0.084982
D                        349  0.060172
E                         22  0.090909

**Вывод:** Среди результатов выделяются самая низкая доля просрочивших среди людей с категорией дохода D (6%) и самая высокая в категории E (9.1%). Однако в целом предположение, что чем больше доход, тем меньше доля просрочивших, не подтверждается.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Прежде всего, посчитаем долю просрочивших для каждой категории целей кредита и общее количество дебиторов в ней:

In [33]:
data.groupby('purpose_category')['debt'].agg(['count', 'mean']).sort_values('mean', ascending=False)

count      mean
purpose_category                         
операции с автомобилем     4279  0.093480
получение образования      3988  0.092528
проведение свадьбы         2313  0.079118
операции с недвижимостью  10751  0.072551

**Вывод:** Меньше всего просрочивших среди тех, кто брал кредит на операции с недвижимостью (7.3%) и проведение свадьбы (7.9%). Несколько выше эта доля среди взявших в кредит на операции с автомобилем или получение образования (9.3%).

<div class="alert-success"> 
<b>Комментарий ревьюера 👍</b>

Везде есть выводы и хороший лаконичный код

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Судя по тому, что пропуски возникают исключительно в двух конкретных столбцах и всегда одновременно, эти пропуски не являются полностью случайными. Нет и причин полагать, что вероятность пропусков зависит от данных в других столбцах, то есть пропуски вряд ли случайные. Вероятно, пропуски неслучайны и некоторые люди намеренно не указали свой заработок и трудовой стаж, сочтя это слишком конфиденциальной информацией или чтобы не снижать шанс на кредит.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍</b>

Согласен с названными тобою возможными причинами пропусков

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* При заполнении пропусков следует использовать наиболее характерные значения относительно имеющихся данных; обычно это среднее или медиана. В данном случае количественные переменные с пропусками - это ежемесячный доход и общий трудовой стаж, то есть величины с большим разбросом в значениях. В этой ситуации взятие среднего приведёт к сдвигу, так как, например, малое количество людей, чей заработок очень велик, сдвинут среднее значение в большую сторону, хотя это никак не будет характеризовать основную часть данных. Для таких переменных, как возраст, где обычно нет такого разброса данных, вполне годится и среднее.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
Верно, дело именно в том, что средние сильно реагируют на выбросы, тем самым искажая наши данные

### Шаг 4: общий вывод.

> В рамках проекта исследовалась статистика платёжеспособности клиентов на основе набора данных из 21525 вхождений, описывающих дебиторов. Работа велась с такими признаками, как количество детей, трудовой стаж, семейное положение, тип занятости, ежемесячный доход, цель получения кредита. Целевым признаком, относительно которого и выявлялись зависимости, был факт наличия задолженности по ходу выплаты кредита.  
  
> В ходе предобработки были обнаружены аномальные отрицательные значения трудового стажа, которые было решено заменить на аналогичные, но положительные. Также из набора были удалены дебиторы с аномальным количеством детей -1 и 20. Пропуски были обнаружены только в столбцах с доходом и стажем, они были заполнены медианными значениями для соответствующего типа занятости. Затем были удалены все дубликаты в количестве 71, а в столбце дохода оставлена лишь целая часть.  В завершение, была проведена категоризация данных по доходу (на 5 категорий с порогами в 30 тыс., 50 тыс., 200 тыс. и 1000 тыс.) и по целям кредита (на 4 категории - операции с недвижимостью, с автомобилями, получение образование, проведение свадьбы).  
  
> Наконец, по имеющимся данным удалось выделить следующие факторы, повышающие риск того, что кредит не будет возвращён в срок: наличие детей, отсутствие опыта брака, доход менее 30000 рублей и направленность кредита на покупку машины или получения образования. Вероятно, именно эти показатели должны учитываться системой скоринга для выявления потенциальных заёмщиков с низкой способностью возвращения кредита.  
  
> В дальнейшем при сборе данных желательно тщательнее подходить к заполнению стажа работы и заработка, чтобы не допускать отрицательных и минимизировать пропущенные значения. Также, в будущем может быть полезно исследовать влияние таких параметров, как образование и пол дебитора, на возврат кредита в срок дл яулучшения системы скоринга.

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 
    
Общий вывод - обязательный и важный элемент работы. Его нужно сформулировать, взяв основные результаты из всех шагов проекта. Начать нужно с того, с какими данными мы работали и какие действия провели на этапе предобработки, а закончить выявленными при ответе на вопросы зависимостями и рекомендациями заказчику по улучшению сбора данных и скоринга. Заказчик, зачастую, читает только общий вывод исследования - поэтому он должен быть максимально информативным

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b>
    
Отличный проект, тобою проделана большая работа, ты молодец!

Сперва похвалю за наличие везде промежуточных выводов, комментариев. Показываешь хорошие навыки владения инструментами Питона, сводными таблицами, group by. Даны верные ответы на поставленные вопросы.
    
Что обязательно стоит доработать: расширить общий вывод. Также я оставил пару желтых комментариев, можешь их тоже исправить
    
Буду ждать доработок :) Удачи!

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 Вторая итерация </b>

Все критические ошибки исправлены, проект стал еще лучше, принимаю. Успехов в дальнейшей учебе!